File: src/lib/server/discovery.ts
import { instructorRatingPrior } from "./instructor-ratings";
import { bayesianRating } from "$lib/instructor-ratings";
import { query } from "./data";
import { gradeSummary } from "$lib/discovery";

export async function coursePreviews(
  items: any[],
  term: string,
  platform?: App.Platform,
  instructor?: string,
) {
  if (!items.length) return [];
  if (items.length > 30) {
    const chunks: any[][] = [];
    for (let i = 0; i < items.length; i += 30)
      chunks.push(
        await coursePreviews(
          items.slice(i, i + 30),
          term,
          platform,
          instructor,
        ),
      );
    return chunks.flat();
  }
  const placeholders = items.map(() => "?").join(",");
  const ids = items.map((row) => row.course_uid);
  const [courses, grades, teachers] = await Promise.all([
    query(
      platform,
      `SELECT uid,payload FROM courses WHERE uid IN (${placeholders})`,
      ids,
    ),
    query(
      platform,
      `SELECT * FROM grade_summaries WHERE uid IN (${placeholders}) AND term<=? AND CAST(term AS INTEGER)>?`,
      [...ids, term, Number(term) - 50],
    ),
    query(
      platform,
      `SELECT DISTINCT t.course_uid,i.uid,i.name,json_extract(i.payload,'$.ratings.quality') quality,json_extract(i.payload,'$.ratings.quality_count') quality_count FROM teaching t JOIN instructors i ON i.uid=t.instructor_uid WHERE t.course_uid IN (${placeholders}) AND t.term=? ORDER BY quality DESC,i.name`,
      [...ids, term],
    ),
  ]);
  const prior = teachers.length ? await instructorRatingPrior(platform) : null;
  const rankedTeachers = teachers.map(row => ({ ...row, quality: bayesianRating(row.quality, row.quality_count ?? 0, prior) })).sort((a, b) => (b.quality ?? -1) - (a.quality ?? -1) || (a.name || "").localeCompare(b.name || ""));
  const payloads = new Map(
    courses.map((row) => [row.uid, JSON.parse(row.payload)]),
  );
  const instructorRows = instructor
    ? await query(
        platform,
        `SELECT uid,term,payload FROM grades WHERE uid IN (${placeholders}) AND section<>'' AND term<=? AND CAST(term AS INTEGER)>? AND EXISTS(SELECT 1 FROM json_each(grades.instructors) WHERE value=?)`,
        [...ids, term, Number(term) - 50, instructor],
      )
    : [];
  return items.map((item) => {
    const course = payloads.get(item.course_uid);
    const claims = [
      ...(course.student_summary?.difficulty_workload || []),
      ...(course.student_summary?.quick_take || []),
    ];
    const teacherRows = instructorRows
      .filter((row) => row.uid === item.course_uid)
      .map((row) => ({ ...JSON.parse(row.payload), term: row.term }));
    const teacherTerms = new Set(teacherRows.map((row) => row.term));
    const teacherClaim = course.student_summary?.current_instructors
      ?.find((row: any) => row.instructor_uid === instructor)
      ?.summary?.find((row: any) => row.citations?.length);
    return {
      ...item,
      description: course.llm_summary?.replace(`${course.course_id} ${course.title} `, "").replace(/^./, (letter: string) => letter.toUpperCase()) || null,
      discovery: {
        term,
        offered: course.offerings.some((row: any) => row.term_id === term),
        history: gradeSummary(
          grades.filter((row) => row.uid === item.course_uid),
        ),
        instructors: rankedTeachers.filter(
          (row) => row.course_uid === item.course_uid,
        ),
        instructorHistory: instructor ? gradeSummary(teacherRows) : null,
        courseComparison: instructor
          ? gradeSummary(
              grades.filter(
                (row) =>
                  row.uid === item.course_uid && teacherTerms.has(row.term),
              ),
            )
          : null,
        claim: instructor
          ? teacherClaim || null
          : claims.find((row: any) =>
              row.citations?.some(
                (citation: any) => citation.type === "review",
              ),
            ) || null,
        reviewFiles: course.evidence?.reviews || [],
      },
    };
  });
}

File: web/uwcourses_site/discovery.py
"""Disposable website indexes for course discovery and original student reviews."""

import json

GRADE_KEYS = ["a", "ab", "b", "bc", "c", "d", "f"]


def build_discovery(db, reviews):
    db.executescript("""
    CREATE TABLE course_numbers(uid TEXT PRIMARY KEY,number INTEGER);
    CREATE TABLE offerings(uid TEXT,term TEXT,PRIMARY KEY(uid,term));
    CREATE TABLE grade_summaries(uid TEXT,term TEXT,a INTEGER,ab INTEGER,b INTEGER,bc INTEGER,c INTEGER,d INTEGER,f INTEGER,PRIMARY KEY(uid,term));
    CREATE TABLE reviews(profile_id TEXT,review_id TEXT,course_uid TEXT,review_date TEXT,payload TEXT,PRIMARY KEY(profile_id,review_id));
    CREATE INDEX reviews_profile_date ON reviews(profile_id,review_date DESC,review_id);
    CREATE INDEX offerings_term ON offerings(term,uid);
    """)
    for uid, payload in db.execute("SELECT uid,payload FROM courses"):
        course = json.loads(payload)
        db.execute(
            "INSERT INTO course_numbers VALUES(?,?)", (uid, course.get("course_number"))
        )
        db.executemany(
            "INSERT OR IGNORE INTO offerings VALUES(?,?)",
            [(uid, row["term_id"]) for row in course["offerings"]],
        )
    columns = ",".join(
        f"COALESCE(json_extract(payload,'$.{key}'),0)" for key in GRADE_KEYS
    )
    db.execute(
        f"INSERT INTO grade_summaries SELECT uid,term,{columns} FROM grades WHERE section='' "
    )
    latest = {}
    for row in reviews:
        key = (row["source_instructor_id"], row["source_review_id"])
        if key not in latest or str(row["observed_at"]) > str(
            latest[key]["observed_at"]
        ):
            latest[key] = row
    db.executemany(
        "INSERT INTO reviews VALUES(?,?,?,?,?)",
        [
            (
                profile,
                review,
                row.get("course_uid"),
                str(row.get("review_date") or ""),
                json.dumps(row, default=str, ensure_ascii=False),
            )
            for (profile, review), row in latest.items()
        ],
    )
